File: 4zadanie.go
package main

import (
	"fmt"
)

func findCandidateIndex(candidates []string, name string) int {
	for i, candidate := range candidates {
		if candidate == name {
			return i
		}
	}

	return -1
}

func getPosition(vote []int, candidate int) int {
	for i, value := range vote {
		if value == candidate {
			return i
		}
	}

	return -1
}

func main() {
	var n int
	var k int

	fmt.Print("Введите количество кандидатов: ")
	fmt.Scan(&n)

	fmt.Print("Введите количество избирателей: ")
	fmt.Scan(&k)

	candidates := make([]string, n)

	fmt.Println("Введите имена кандидатов:")

	for i := range candidates {
		fmt.Scan(&candidates[i])
	}

	votes := make([][]int, k)

	fmt.Println("Введите предпочтения каждого избирателя от лучшего к худшему:")

	for i := 0; i < k; i++ {
		votes[i] = make([]int, n)

		fmt.Printf("Избиратель %d: ", i+1)

		for j := 0; j < n; j++ {
			var name string
			fmt.Scan(&name)

			index := findCandidateIndex(candidates, name)

			if index == -1 {
				fmt.Println("Ошибка: кандидат не найден:", name)
				return
			}

			votes[i][j] = index
		}
	}

	// Метод Борда
	bordaScores := make([]int, n)

	for _, vote := range votes {
		for position, candidate := range vote {
			bordaScores[candidate] += n - 1 - position
		}
	}

	bordaWinner := 0

	for i := 1; i < n; i++ {
		if bordaScores[i] > bordaScores[bordaWinner] {
			bordaWinner = i
		}
	}

	fmt.Println("\nРезультаты по методу Борда:")

	for i := 0; i < n; i++ {
		fmt.Printf("%s: %d баллов\n", candidates[i], bordaScores[i])
	}

	fmt.Println("Победитель по Борду:", candidates[bordaWinner])

	// Метод Кондорсе
	condorcetWinner := -1

	for candidate := 0; candidate < n; candidate++ {
		winsAgainstAll := true

		for opponent := 0; opponent < n; opponent++ {
			if candidate == opponent {
				continue
			}

			candidateWins := 0
			opponentWins := 0

			for _, vote := range votes {
				candidatePosition := getPosition(vote, candidate)
				opponentPosition := getPosition(vote, opponent)

				if candidatePosition < opponentPosition {
					candidateWins++
				} else {
					opponentWins++
				}
			}

			if candidateWins <= opponentWins {
				winsAgainstAll = false
				break
			}
		}

		if winsAgainstAll {
			condorcetWinner = candidate
			break
		}
	}

	fmt.Println("\nРезультат по методу Кондорсе:")

	if condorcetWinner == -1 {
		fmt.Println("Победитель по Кондорсе не определен")
	} else {
		fmt.Println("Победитель по Кондорсе:", candidates[condorcetWinner])
	}

	if condorcetWinner != -1 && condorcetWinner != bordaWinner {
		fmt.Println("\nМетоды дали разных победителей.")
	}

	if condorcetWinner == -1 {
		fmt.Println("\nПо методу Кондорсе победитель может не существовать, если нет кандидата, который выигрывает у всех остальных в парных сравнениях.")
	}
}

File: 4zadanie.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

int findCandidateIndex(const vector<string>& candidates, const string& name) {
    for (size_t i = 0; i < candidates.size(); i++) {
        if (candidates[i] == name) {
            return static_cast<int>(i);
        }
    }

    return -1;
}

int getPosition(const vector<int>& vote, int candidate) {
    for (size_t i = 0; i < vote.size(); i++) {
        if (vote[i] == candidate) {
            return static_cast<int>(i);
        }
    }

    return -1;
}

int main() {
    int n;
    cout << "Введите количество кандидатов: ";
    cin >> n;

    int nIsbirateley;
    cout << "Введите количество избирателей: ";
    cin >> nIsbirateley;

    vector<string> candidates(n);
    cout << "Введите имена кандидатов:\n";
    for (string& candidate : candidates) {
        cin >> candidate;
    }

    vector<vector<int>> votes(nIsbirateley, vector<int>(n));
    cout << "Введите предпочтения каждого избирателя от лучшего к худшему:\n";
    for (int i = 0; i < nIsbirateley; i++) {
        cout << "Избиратель " << i + 1 << ": ";

        for (int j = 0; j < n; j++) {
            string name;
            cin >> name;

            votes[i][j] = findCandidateIndex(candidates, name);
        }
    }

    vector<int> bordaScores(n, 0);

    for (const vector<int>& vote : votes) {
        for (int position = 0; position < n; position++) {
            int candidate = vote[position];
            bordaScores[candidate] += n - 1 - position;
        }
    }

    int bordaWinner = 0;

    for (int i = 1; i < n; i++) {
        if (bordaScores[i] > bordaScores[bordaWinner]) {
            bordaWinner = i;
        }
    }

    cout << "\nРезультаты по методу Борда:\n";

    for (int i = 0; i < n; i++) {
        cout << candidates[i] << ": " << bordaScores[i] << " баллов\n";
    }

    cout << "Победитель по Борду: " << candidates[bordaWinner] << '\n';

    int condorcetWinner = -1;

    for (int candidate = 0; candidate < n; candidate++) {
        bool winsAgainstAll = true;

        for (int opponent = 0; opponent < n; opponent++) {
            if (candidate == opponent) {
                continue;
            }

            int candidateWins = 0;
            int opponentWins = 0;

            for (const vector<int>& vote : votes) {
                int candidatePosition = getPosition(vote, candidate);
                int opponentPosition = getPosition(vote, opponent);

                if (candidatePosition < opponentPosition) {
                    candidateWins++;
                } else {
                    opponentWins++;
                }
            }
            
            if (candidateWins <= opponentWins) {
                winsAgainstAll = false;
                break;
            }
        }

        if (winsAgainstAll) {
            condorcetWinner = candidate;
            break;
        }
    }

    cout << "\nРезультат по методу Кондорсе:\n";

    if (condorcetWinner == -1) {
        cout << "Победитель по Кондорсе не определен\n";
    } else  {
        cout << "Победитель по Кондорсе: " << candidates[condorcetWinner] << '\n';
    }

    if (condorcetWinner != -1 && condorcetWinner != bordaWinner) {
        cout << "\nМетоды дали разных победителей.\n";
    }

    return 0;
}